File: remember/scheduler.py
"""
Automatic archival scheduler for remember-mcp
Runs background task to archive old/decayed memories
"""
from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, Optional

logger = logging.getLogger(__name__)


class ArchivalScheduler:
    """
    Background scheduler for automatic memory archival.

    Periodically checks for memories eligible for archival based on:
    - Age (days since creation)
    - Salience (decay level)
    - Last access time
    """

    def __init__(
        self,
        remember_system,
        interval_seconds: int = 86400,  # Default: 24 hours
        enabled: bool = True
    ):
        """
        Initialize archival scheduler.

        Args:
            remember_system: RememberSystem instance
            interval_seconds: Interval between archival checks (default: 24h)
            enabled: Whether scheduler is enabled
        """
        if interval_seconds < 1:
            raise ValueError("interval_seconds must be >= 1")
        self.system = remember_system
        self.interval = interval_seconds
        self.enabled = enabled
        self.running = False
        self.task: Optional[asyncio.Task] = None
        self.last_run: Optional[int] = None
        self.total_archived = 0
        self.last_error: Optional[str] = None

    async def start(self) -> None:
        """Start the scheduler"""
        if self.running:
            logger.info("Scheduler already running")
            return

        self.enabled = True
        self.running = True
        self.task = asyncio.create_task(self._run_loop())
        logger.info("Scheduler started (interval: %ss)", self.interval)

    async def stop(self) -> None:
        """Stop the scheduler"""
        if not self.running:
            return

        self.running = False
        if self.task:
            self.task.cancel()
            try:
                await self.task
            except asyncio.CancelledError:
                pass
            self.task = None

        logger.info("Scheduler stopped")

    async def _run_loop(self) -> None:
        """Main scheduler loop"""
        while self.running:
            try:
                await self._run_archival()
                await asyncio.sleep(self.interval)
            except asyncio.CancelledError:
                break
            except Exception:
                logger.exception("Scheduler loop error")
                await asyncio.sleep(60)  # Wait 1 minute before retry

    async def _run_archival(self) -> None:
        """Run archival process"""
        if not self.enabled:
            return

        start_time = time.time()
        logger.info("Running archival")

        try:
            stats = await self.system.archive_old_memories(
                age_days=self.system.archive_threshold_days,
                min_salience=self.system.archive_min_salience
            )

            self.last_run = int(time.time() * 1000)
            self.total_archived += stats.archived_count
            self.last_error = None

            elapsed = time.time() - start_time

            if stats.archived_count > 0:
                logger.info(
                    "Archived %s memories (%s bytes, %.2fx) in %.2fs; lifetime total %s",
                    stats.archived_count,
                    stats.archive_size_bytes,
                    stats.compression_ratio,
                    elapsed,
                    self.total_archived,
                )
            else:
                logger.info(
                    "No memories eligible for archival (%.2fs)", elapsed
                )

        except Exception as exc:
            self.last_error = str(exc)
            logger.exception("Archival failed")

    async def run_now(self) -> None:
        """Manually trigger archival immediately"""
        logger.info("Manual archival triggered")
        await self._run_archival()

    def get_status(self) -> dict[str, Any]:
        """Get scheduler status"""
        return {
            "running": self.running,
            "enabled": self.enabled,
            "interval_seconds": self.interval,
            "last_run": self.last_run,
            "total_archived": self.total_archived,
            "next_run_in": self._time_until_next_run(),
            "last_error": self.last_error,
        }

    def _time_until_next_run(self) -> Optional[int]:
        """Calculate seconds until next run"""
        if not self.running or self.last_run is None:
            return None

        elapsed = int(time.time() * 1000) - self.last_run
        remaining = (self.interval * 1000) - elapsed

        return max(0, remaining // 1000)

File: src/remember/scheduler.ts
/**
 * Background scheduler for automatic memory archival.
 */

import type { RememberSystem } from "./system.ts";

export class ArchivalScheduler {
  readonly system: RememberSystem;
  interval: number;
  enabled: boolean;
  running = false;
  private timer: ReturnType<typeof setTimeout> | null = null;
  last_run: number | null = null;
  total_archived = 0;
  last_error: string | null = null;

  constructor(
    rememberSystem: RememberSystem,
    intervalSeconds = 86_400,
    enabled = true,
  ) {
    if (intervalSeconds < 1) throw new Error("interval_seconds must be >= 1");
    this.system = rememberSystem;
    this.interval = intervalSeconds;
    this.enabled = enabled;
  }

  async start(): Promise<void> {
    if (this.running) return;
    this.enabled = true;
    this.running = true;
    this._scheduleNext(0);
  }

  async stop(): Promise<void> {
    this.running = false;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }

  private _scheduleNext(delayMs: number): void {
    if (!this.running) return;
    this.timer = setTimeout(() => {
      void this._tick();
    }, delayMs);
  }

  private async _tick(): Promise<void> {
    if (!this.running) return;
    try {
      await this._runArchival();
      this._scheduleNext(this.interval * 1000);
    } catch (err) {
      this.last_error = err instanceof Error ? err.message : String(err);
      this._scheduleNext(60_000);
    }
  }

  private async _runArchival(): Promise<void> {
    if (!this.enabled) return;
    try {
      const stats = await this.system.archiveOldMemories({
        age_days: this.system.archiveThresholdDays,
        min_salience: this.system.archiveMinSalience,
      });
      this.last_run = Date.now();
      this.total_archived += stats.archived_count;
      this.last_error = null;
    } catch (err) {
      this.last_error = err instanceof Error ? err.message : String(err);
      throw err;
    }
  }

  async runNow(): Promise<void> {
    await this._runArchival();
  }

  getStatus(): Record<string, unknown> {
    return {
      running: this.running,
      enabled: this.enabled,
      interval_seconds: this.interval,
      last_run: this.last_run,
      total_archived: this.total_archived,
      next_run_in: this._timeUntilNextRun(),
      last_error: this.last_error,
    };
  }

  private _timeUntilNextRun(): number | null {
    if (!this.running || this.last_run === null) return null;
    const elapsed = Date.now() - this.last_run;
    const remaining = this.interval * 1000 - elapsed;
    return Math.max(0, Math.floor(remaining / 1000));
  }
}
